src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java

```diff
@@ -222,60 +239,167 @@ private void endStream(RowStream<Row> stream, RoutingContext ctx,
             return con.preparedQuery("INSERT INTO " + titleEntriesTable(pool)
                 + "(id, counterReportTitle, matchCriteria,"
                 + " kbTitleName, kbTitleId,"
                 + " kbManualMatch, printISSN, onlineISSN)"
                 + " VALUES ($1, $2, $3, $4, $5, $6, $7, $8)"
-                + " ON CONFLICT (counterReportTitle) DO NOTHING")
-                .execute(Tuple.of(UUID.randomUUID(), counterReportTitle, onlineIssn,
+                + " ON CONFLICT (matchCriteria) DO NOTHING")
+                .execute(Tuple.of(UUID.randomUUID(), counterReportTitle, match,
                     kbTitleName, kbTitleId,
                     false, printIssn, onlineIssn))
                 .compose(x ->
                     con.preparedQuery("SELECT id FROM " + titleEntriesTable(pool)
-                        + " WHERE counterReportTitle = $1")
-                        .execute(Tuple.of(counterReportTitle))
+                        + " WHERE matchCriteria = $1")
+                        .execute(Tuple.of(match))
                         .map(res2 -> res2.iterator().next().getUUID(0))
                 );
           });
         });
   }
 
+  Future<Void> updateTitleFromAgreement(TenantPgPool pool, SqlConnection con,
+                                        UUID kbTitleId, RoutingContext ctx) {
+    if (kbTitleId == null) {
+      return Future.succeededFuture();
+    }
+    return con.preparedQuery("SELECT * FROM " + titleEntriesTable(pool)
+        + " WHERE kbTitleId = $1")
+        .execute(Tuple.of(kbTitleId))
+        .compose(res -> {
+          if (!res.iterator().hasNext()) {
+            return Future.succeededFuture();
+          }
+          return ermTitleLookup(ctx, kbTitleId).compose(erm -> {
+            String kbTitleName = erm.getString(1);
+            String match = "online:" + erm.getString(2);
+            return con.preparedQuery("INSERT INTO " + titleEntriesTable(pool)
+                + "(id, matchCriteria, kbTitleName, kbTitleId,"
+                + " kbManualMatch)"
+                + " VALUES ($1, $2, $3, $4, $5)"
+                + " ON CONFLICT (matchCriteria) DO NOTHING")
+                .execute(Tuple.of(UUID.randomUUID(), match, kbTitleName, kbTitleId, false))
+                .mapEmpty();
+          });
+        });
+  }
+
+  Future<Void> updatePackageFromAgreement(TenantPgPool pool, SqlConnection con, UUID kbPackageId,
+                                          String kbPackageName, RoutingContext ctx) {
+    if (kbPackageId == null) {
+      return Future.succeededFuture();
+    }
+    return con.preparedQuery("SELECT * FROM " + packageEntriesTable(pool)
+        + " WHERE kbPackageId = $1")
+        .execute(Tuple.of(kbPackageId))
+        .compose(res -> {
+          if (res.iterator().hasNext()) {
+            return Future.succeededFuture();
+          }
+          return ermPackageContentLookup(ctx, kbPackageId)
+              .compose(list -> {
```

Comment:
This deeply nested compose can be avoided:

```
          return ermPackageContentLookup(ctx, kbPackageId);
        })
        .compose(list -> {
```

Reply:
Not really.